core/src/apps/cardano/helpers/tx_builder.py

```diff
@@ -0,0 +1,194 @@
+    from trezor.messages.CardanoToken import CardanoToken
+    from trezor.messages.CardanoTxCertificate import CardanoTxCertificate
+    from trezor.messages.CardanoTxWithdrawal import CardanoTxWithdrawal
+    from .. import seed
+    from typing import Any, Optional
+
+INPUT = const(0)
+OUTPUT = const(1)
+ASSET_GROUP = const(2)
+TOKEN = const(3)
+CERTIFICATE = const(4)
+WITHDRAWAL = const(5)
+
+
+class TxBuilder:
+    def __init__(self, keychain: seed.Keychain, msg: CardanoSignTx):
+        tx_body_map_item_count = 3 + sum(
+            1
+            for entry in [
+                msg.ttl,
+                msg.certificates_count,
+                msg.withdrawals_count,
+                msg.has_auxiliary_data,
+                msg.validity_interval_start,
+            ]
+            if bool(entry)
+        )
+        self.state = [LazyCollection(DICT, tx_body_map_item_count)]
+        self.serializer = cbor.encode_streamed(self.state[0])
+        self.hashfn = hashlib.blake2b(outlen=32)
+        self._update_tx_hash()
+        self.keychain = keychain
+        self.msg = msg
+        self.expected_item_kind: Optional[int] = None
+
+    def _update_tx_hash(self) -> None:
+        while (chunk := next(self.serializer)) is not None:
+            self.hashfn.update(chunk)
+
+    def get_tx_hash(self) -> bytes:
+        assert self.state[0].is_filled(), "The hash calculation is not finished yet"
+        return self.hashfn.digest()
+
+    def _start_collection_at_key(
+        self,
+        key: Any,
+        collection: LazyCollection,
+        next_kind: int,
+        current_kind: Optional[int] = None,
+    ) -> None:
+        self._add_item((key, collection), current_kind)
+        self.state.append(collection)
+        self.expected_item_kind = next_kind
+
+    def _add_item(self, item: Any, item_kind: Optional[int] = None) -> None:
+        assert self.expected_item_kind == item_kind, "Item of unexpected kind added"
+        self.state[-1].append_item(item)
+        self._update_tx_hash()
+
+    def _finish_collection(
+        self, current_kind: int, next_kind: Optional[int] = None
+    ) -> None:
+        assert (
+            self.expected_item_kind == current_kind
+        ), "Tried to finish a collection of unexpected kind"
+        self.expected_item_kind = next_kind
+        assert self.state[-1].is_filled(), "The collection is not filled yet"
+        self.state.pop()
+        assert next(self.serializer) is None
+
+    def start_inputs(self) -> None:
+        self._start_collection_at_key(
+            0, LazyCollection(LIST, self.msg.inputs_count), INPUT
+        )
+
+    def add_input(self, input: CardanoTxInput) -> None:
+        self._add_item((input.prev_hash, input.prev_index), INPUT)
+
+    def finish_inputs(self) -> None:
+        self._finish_collection(INPUT)
+
+    def start_outputs(self) -> None:
+        self._start_collection_at_key(
+            1, LazyCollection(LIST, self.msg.outputs_count), OUTPUT
```

Comment:
I'd consider passing the parameters as named ones, so the reader doesn't have to look up the function definition and also so it's resilient against function parameter reorganization/renaming in the future